editoast/src/views/scenario/macro_nodes.rs

```diff
@@ -0,0 +1,614 @@
+use axum::extract::Json;
+use axum::extract::Path;
+use axum::extract::Query;
+use axum::extract::State;
+use axum::http::StatusCode;
+use axum::response::IntoResponse;
+use axum::Extension;
+use diesel_async::scoped_futures::ScopedFutureExt;
+use editoast_authz::BuiltinRole;
+use editoast_derive::EditoastError;
+use editoast_models::DbConnection;
+use editoast_models::DbConnectionPoolV2;
+use itertools::Itertools;
+use serde::Deserialize;
+use serde::Serialize;
+use thiserror::Error;
+use utoipa::IntoParams;
+use utoipa::ToSchema;
+
+use crate::error::InternalError;
+use crate::error::Result;
+use crate::models::macro_node::MacroNode;
+use crate::models::prelude::*;
+use crate::models::Project;
+use crate::models::Scenario;
+use crate::models::Study;
+use crate::views::pagination::PaginatedList;
+use crate::views::pagination::PaginationQueryParam;
+use crate::views::pagination::PaginationStats;
+use crate::views::projects::ProjectError;
+use crate::views::projects::ProjectIdParam;
+use crate::views::scenario::ScenarioError;
+use crate::views::scenario::ScenarioIdParam;
+use crate::views::study::StudyError;
+use crate::views::study::StudyIdParam;
+use crate::views::AuthenticationExt;
+use crate::views::AuthorizationError;
+
+crate::routes! {
+    "/macro_nodes" => {list, create,},
+    "/macro_nodes/{node_id}" => {get, update, delete,},
+}
+
+#[derive(Debug, Error, EditoastError)]
+#[editoast_error(base_id = "macro_node")]
+pub enum MacroNodeError {
+    #[error("Node '{node_id}', could not be found")]
+    #[editoast_error(status = 404)]
+    NotFound { node_id: i64 },
+}
+
+#[derive(IntoParams, Deserialize)]
+#[allow(unused)]
+pub struct MacroNodeIdParam {
+    node_id: i64,
+}
+
+#[derive(Debug, Serialize, Deserialize, PartialEq, Clone, ToSchema)]
+pub struct MacroNodeForm {
+    pub position_x: f64,
+    pub position_y: f64,
+    pub full_name: Option<String>,
+    pub connection_time: i64,
+    pub labels: Vec<Option<String>>,
+    pub trigram: Option<String>,
+    pub path_item_key: String,
+}
+
+impl MacroNodeForm {
+    pub fn into_macro_node_changeset(self, scenario_id: i64) -> Result<Changeset<MacroNode>> {
+        let macro_node = MacroNode::changeset()
+            .scenario_id(scenario_id)
+            .position_x(self.position_x)
+            .position_y(self.position_y)
+            .full_name(self.full_name)
+            .connection_time(self.connection_time)
+            .labels(self.labels)
+            .trigram(self.trigram)
+            .path_item_key(self.path_item_key);
+
+        Ok(macro_node)
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize, PartialEq, Clone, ToSchema)]
```

Comment:
```suggestion
#[derive(Debug, Serialize, ToSchema)]
#[cfg_attr(test, derive(Deserialize, PartialEq))]
```

We only need to derive these for testing.